### PR TITLE
SESE: deal with the case where outer loops are moved into the current loop.

### PR DESCRIPTION
**Pull request by bgogul**
Consider the following example:
```
while ... {
  if ... {
     for(...) {...} // This should be nested into the while loop.
     break;
  }
}
```
In the SIL CFG, the statement in the `if` block including the `for` loop will not be part of the outer `while` loop. One of the steps during SESE canonicalization is to transform the CFG so that the statements in the `if` block become part of the `while` loop. This PR takes care of this case by doing the following:
  - Canonicalizing the `for` loop first.
  - Nesting the canonicalized `for` loop into the `while` loop and updating LoopInfo as needed.

This also takes care of the case with loops analogous to the situation handled in https://github.com/apple/swift/pull/19570 (See the example added to sese_loop_canonicalization.sil in this PR.)

This is the final PR for https://bugs.swift.org/browse/SR-7765 (modulo unknown bugs).

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 1]
```diff
@@ -771,24 +842,45 @@ void SingleExitLoopTransformer::ensureSingleExitBlock() {
...
       //   }
       // }
-      // Check that `loop` is nested within `reachableLoop`.
-      assert(outsideBlockLoop->contains(loop) &&
-             "Nodes being moved belong to a non-nested loop.");
-      // Move the node into our loop.
-      outsideBlockLoop->removeBlockFromLoop(outsideBlock);
-      LI->changeLoopFor(outsideBlock, nullptr);
+      if (outsideBlockLoop->contains(loop)) {
+        // If our `loop` is nested within `outsideBlockLoop`.  Move the node
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: is there test coverage on both this if branch, and the else below?

**Review by mhong — APPROVED** [review]
Congrats on completing the feature and bug fix work!

(There's still some clean-up work remaining, but pls feel free to resolve the bug first if you prefer.)

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 2]
```diff
@@ -514,24 +514,46 @@ class BasicBlockCloner : public SILClonerWithScopes<BasicBlockCloner> {
...
-                            PostDominanceInfo *PDI)
-      : deviceInfo(deviceInfo), DI(DI), PDI(PDI), LI(LI), loop(loop),
-        header(loop->getHeader()), preheader(loop->getLoopPreheader()),
-        latch(loop->getLoopLatch()), currentFn(header->getParent()),
-        oldHeaderNumArgs(header->getNumArguments()), hasUndefsAtPreheader(false) {
-    assert(preheader && "Canonicalization should have given us one preheader");
-    assert(latch && "Canonicalization should have given us one latch block");
-    initialize();
+  public:
+    static bool doIt(GraphFunctionDeviceInfo *deviceInfo, SILLoopInfo *LI,
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 3]
```diff
@@ -711,6 +733,24 @@ void SingleExitLoopTransformer::ensureSingleExitBlock() {
...
+  // from outside is moved into the current loop. e.g.,
+  //   while ... {
+  //     if ... {
+  //       for(...) {...} // This should be nested into the while loop.
+  //       break;
+  //     }
+  //   }
+  SmallPtrSet<SILBasicBlock *, 32> unrelatedPreheaders;
+  for (auto *otherLoop : *LI) {
+    // If loop and otherLoop are not nested into either, remember the preheader.
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 4]
```diff
@@ -752,6 +805,24 @@ void SingleExitLoopTransformer::ensureSingleExitBlock() {
 
         // Clone the block and rewire the edge.
         SILBasicBlock *clonedSucc = cloner.initAndCloneBlock(succ);
+        // If `succ` is a preheader of an unrelated loop, we will have to clone
+        // the entire loop now so that we can also incrementally update LoopInfo.
+        if (succBlockLoop) {
+          SILLoop *clonedLoop =
+              cloner.cloneLoop(LI, succBlockLoop, succBlockLoop->getHeader());
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Done and updated the comment as well.

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp (hunk 1 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Yes, we have coverage for both. The `if` branch was handled even before this PR (and is covered by the example with labeled breaks).

The else branch is the new behavior and the tests in this PR provide coverage.

File: test/TensorFlow/sese_loop_canonicalization.sil [hunk 5]
```diff
@@ -361,3 +361,88 @@ bb8:
 bb9(%18 : $Builtin.Int32):
   return %18 : $Builtin.Int32
 }
+
+// The following example is similar to loopThatRequiresNodeCloning, where the
```
[the remark this answers is not in the document]
Comment (bgogul) [reply test/TensorFlow/sese_loop_canonicalization.sil]: This is representing a case where we need to move an outside loop into the currently processed loop. However, the header of the outside loop is not dominated by the currently processed loop. In this case, we will have to clone the loop and make sure the LoopInfo is updated. 

We want a very specific CFG to test this scenario. See the ASCII graph in the test file for the CFG. With a swift file, the generated CFG can change depending upon the version of the compiler. (Also, the compiler did generate a similar CFG for an example.)

File: test/TensorFlow/sese_loop_canonicalization.swift [hunk 6]
```diff
@@ -5,6 +5,54 @@
...
 import TensorFlow
 
+//expected-warning @+1 {{value implicitly copied to the host}}
+public func testLoopMovementFromOutside(_ breakIndex: Int32) -> Tensor<Int32> {
+  var i: Int32 = 1
+  var sum = Tensor<Int32>(0)
+  let maxCount: Int32 = 100
+  while i <= maxCount {
+    sum += i
+    if (i == breakIndex) {
```
[the remark this answers is not in the document]
Comment (bgogul) [reply test/TensorFlow/sese_loop_canonicalization.swift]: done

File: test/TensorFlow/sese_loop_canonicalization.swift [hunk 7]
```diff
@@ -5,6 +5,54 @@
...
+      break // expected-note {{value used here}}
+    }
+    i += 1
+  }
+  return sum
+}
+
+// CHECK-LABEL: --- XLA CFG Loops Before Canonicalize: {{.*}}testLoopMovementFromOutside{{.*}}
+// CHECK: Loop at depth 1 containing: {{.*}}
+// CHECK: Loop at depth 1 containing: {{.*}}
```
[the remark this answers is not in the document]
Comment (bgogul) [reply test/TensorFlow/sese_loop_canonicalization.swift]: Done.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by bgogul** [comment]
Thanks for the review, Ming. Merging it now.

**Pull request merged**